### PR TITLE
chore: migrate SettlementButton.js class to function component

### PR DESCRIPTION
**Pull request by teneeto**
### Details

This change is made to migrate the [SettlementButton.js  ](https://github.com/Expensify/App/blob/main/src/components/SettlementButton.js) from class to a function component

### Fixed Issues

$: [16201](https://github.com/Expensify/App/issues/16201)
PROPOSAL:  [16201](https://github.com/Expensify/App/issues/16201)

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/22779249/fc4599e6-4f6c-42dd-8ba7-1a5d06db93f7

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/22779249/d0fed7a0-4cb2-452f-851b-8be86e523730

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/22779249/e513d57e-7d17-43be-b3e9-a2450d2d6f6e

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/22779249/ceb258f3-ad49-407e-bf22-baeaeec6b210

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/22779249/ea41c4ed-d8d8-4332-a02b-cfb296c4beb5

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/22779249/8778cf3a-57b0-4313-98e2-d1570bbfaf67

</details>

### CODE REVIEW

File: src/components/SettlementButton.js [hunk 1]
```diff
@@ -60,17 +58,31 @@ const propTypes = {
...
     buttonSize: PropTypes.oneOf(_.values(CONST.DROPDOWN_BUTTON_SIZE)),
 
-    /** The anchor alignment of the popover menu */
+    /** Route for the Add Bank Account screen for a given navigation stack */
+    addBankAccountRoute: PropTypes.string,
+
+    /** Route for the Add Debit Card screen for a given navigation stack */
+    addDebitCardRoute: PropTypes.string,
+
+    /** Should the confirmation button be disabled? */
```
Comment (rezkiy37) [inline src/components/SettlementButton.js:67]:
```suggestion
    /** Whether the button should be disabled */
```

File: src/components/SettlementButton.js [hunk 2]
```diff
@@ -88,43 +100,64 @@ const defaultProps = {
...
 
-    getButtonOptionsFromProps() {
+    const getButtonOptionsFromProps = () => {
         const buttonOptions = [];
-        const isExpenseReport = ReportUtils.isExpenseReport(this.props.iouReport);
+        const isExpenseReport = ReportUtils.isExpenseReport(iouReport);
         const paymentMethods = {
             [CONST.IOU.PAYMENT_TYPE.EXPENSIFY]: {
-                text: this.props.translate('iou.settleExpensify', {formattedAmount: this.props.formattedAmount || ''}),
+                text: translate('iou.settleExpensify', {formattedAmount: formattedAmount || ''}),
```
Comment (rezkiy37) [inline src/components/SettlementButton.js:135]: `formattedAmount` is a props, that has a default value `""`. Do we really need have one more default value for this?
```suggestion
                text: translate('iou.settleExpensify', {formattedAmount}),
```

File: src/components/SettlementButton.js [hunk 3]
```diff
@@ -168,48 +201,45 @@ class SettlementButton extends React.Component {
...
-                        buttonSize={this.props.buttonSize}
-                        anchorAlignment={this.props.anchorAlignment}
-                    />
-                )}
-            </KYCWall>
-        );
-    }
+    return (
+        <KYCWall
+            onSuccessfulKYC={(iouPaymentType) => onPress(iouPaymentType)}
```
Comment (rezkiy37) [inline src/components/SettlementButton.js:206]:
```suggestion
            onSuccessfulKYC={onPress}
```

File: src/components/SettlementButton.js [hunk 4]
```diff
@@ -168,48 +201,45 @@ class SettlementButton extends React.Component {
...
+            isDisabled={isOffline}
+            chatReportID={chatReportID}
+            iouReport={iouReport}
+        >
+            {(triggerKYCFlow, buttonRef) => (
+                <ButtonWithDropdownMenu
+                    buttonRef={buttonRef}
+                    isDisabled={isDisabled}
+                    isLoading={isLoading}
+                    onPress={(event, iouPaymentType) => {
```
Comment (rezkiy37) [inline src/components/SettlementButton.js:219]: Could we create a handler, because it isn't an one-line callback.

**Review by rezkiy37 — COMMENTED** [review]
Left comments.

File: src/components/SettlementButton.js [hunk 5]
```diff
@@ -88,43 +100,64 @@ const defaultProps = {
...
+    const {translate} = useLocalize();
+    const {isOffline} = useNetwork();
+
+    useEffect(() => {
         PaymentMethods.openWalletPage();
-    }
+    }, []);
 
-    getButtonOptionsFromProps() {
+    const getButtonOptionsFromProps = () => {
```
Comment (gedu) [inline src/components/SettlementButton.js:130]: I'm thinking that this should be memoized, we only care about the end value, and there are a lot of stuff going on here. 
If we memoize there are a couple of things we need to tackle, `betas` and `nvp_lastPaymentMethod` aren't stable default values. I would update the default values with stable ones like
```
const EMPTY_ARRAY = Object.freeze([]);
const EMPTY_OBJECT = Object.freeze({});

 betas: EMPTY_ARRAY,
 nvp_lastPaymentMethod: EMPTY_OBJECT,
 ```
First I will check how many times this component re-renders, to check if worth the change.
Then, I would double-check what Onyx returns, maybe we can make it stable too

File: src/components/SettlementButton.js [hunk 6]
```diff
@@ -1,21 +1,29 @@
+import React, {useEffect, useMemo} from 'react';
 import _ from 'underscore';
-import React from 'react';
 import PropTypes from 'prop-types';
 import {withOnyx} from 'react-native-onyx';
-import ButtonWithDropdownMenu from './ButtonWithDropdownMenu';
-import * as Expensicons from './Icon/Expensicons';
-import Permissions from '../libs/Permissions';
-import ONYXKEYS from '../ONYXKEYS';
+
```
Comment (BeeMargarida) [inline src/components/SettlementButton.js:5]: I also like this way of organizing imports, but it seems like the standard here is to not have empty lines between imports

File: src/components/SettlementButton.js [hunk 7]
```diff
@@ -88,43 +106,64 @@ const defaultProps = {
...
+    const {translate} = useLocalize();
+    const {isOffline} = useNetwork();
+
+    useEffect(() => {
         PaymentMethods.openWalletPage();
-    }
+    }, []);
 
-    getButtonOptionsFromProps() {
+    const getButtonOptionsFromProps = useMemo(() => {
```
Comment (gedu) [inline src/components/SettlementButton.js:136]: Update the name, isn't a function now. It should be something like `paymentOptions` or `paymentButtonOptions`

File: src/CONST.ts [hunk 8]
```diff
@@ -4,6 +4,11 @@ import * as KeyCommand from 'react-native-key-command';
 import * as Url from './libs/Url';
 import SCREENS from './SCREENS';
 
+// Creating a default array and object this way because objects ({}) and arrays ([]) are not stable types.
+// Freezing the array ensures that it cannot be unintentionally modified.
+const EMPTY_ARRAY = Object.freeze([]);
+const EMPTY_OBJECT = Object.freeze({});
```
Comment (amyevans) [inline src/CONST.ts:10]: I see similar definitions in https://github.com/Expensify/App/blob/main/src/hooks/useArrowKeyFocusManager.js and https://github.com/Expensify/App/blob/main/src/hooks/useKeyboardShortcut.js as well, let's clean those up to use these new constants.

Could you also add a comment explaining where/why this is useful?

**Review by amyevans — CHANGES_REQUESTED** [review]
Just some minor feedback, otherwise looks good!

**Review by amyevans — DISMISSED** [review]
Thanks!

Going to hold merge until Monday given the freeze.

**Comment by amyevans** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by teneeto** [comment]
> Thanks!
> 
> Going to hold merge until Monday given the freeze.

Saw the lint failed, so fixed the prettier diff.

**Review by amyevans — APPROVED** [review]

**Pull request merged**